scheduler/filter/utils.go

```diff
@@ -0,0 +1,92 @@
+package filter
+
+import (
+	"regexp"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type comparison int
+
+const (
+	equ = comparison(iota)
+	neg
+	gte
+	lte
+)
+
+func parse(k, v string) (string, string, comparison, bool) {
+	// default comparison mode
+	mode := equ
+
+	// support case of constraint:k==v
+	// with "=", it's possible for an entry "k==v" to be split as:
+	// 1. "k=" as key and "v" as value
+	// 2. "k" as key and "=v" as value
+	// Just to make sure it cover these cases.
+	if strings.HasSuffix(k, "=") {
+		k = strings.TrimSuffix(k, "=")
+	} else if strings.HasPrefix(v, "=") {
+		v = strings.TrimPrefix(v, "=")
+	}
+
+	if strings.HasPrefix(v, "!") {
+		log.Debugf("negate detected in value")
+		v = strings.TrimPrefix(v, "!")
+		mode = neg
+	} else if strings.HasSuffix(k, "!") {
+		log.Debugf("negate detected in key")
+		k = strings.TrimSuffix(k, "!")
+		mode = neg
+	} else {
+		if strings.HasSuffix(k, ">") {
+			log.Debugf("gt (>) detected in key")
+			k = strings.TrimSuffix(k, ">")
+			mode = gte
+		} else if strings.HasSuffix(k, "<") {
+			log.Debugf("lt (<) detected in key")
+			k = strings.TrimSuffix(k, "<")
+			mode = lte
+		}
+	}
+
+	useRegex := false
+	if strings.HasPrefix(v, "/") && strings.HasSuffix(v, "/") {
+		log.Debugf("regex detected")
+		v = strings.TrimPrefix(strings.TrimSuffix(v, "/"), "/")
+		useRegex = true
+	}
+
+	return k, v, mode, useRegex
+}
+
+func extractEnv(key string, env []string) map[string]string {
+	values := make(map[string]string)
+	for _, e := range env {
+		if strings.HasPrefix(e, key+":") {
+			value := strings.TrimPrefix(e, key+":")
+			parts := strings.SplitN(value, "=", 2)
+			if len(parts) == 2 {
+				values[strings.ToLower(parts[0])] = strings.ToLower(parts[1])
```

Comment:
Not sure if `ToLower` (at least, on the *value*) is the right thing to do here. I'd keep the value as-is and decide if matches should be case-insensitive during the actual filtering.

Perhaps white-space  *should* be trimmed from value and key to allow `something == something`?